### ISSUE TITLE
How might I better leverage GPUs?

### ISSUE THREAD

**Issue by mbuckler** [facebookresearch/rebel#12]
Thanks for releasing the code! 

TLDR: The default configuration under-utilizes my GPUs but I don't know how to change this. How can I better utilized my GPUs?

Details:
I have a desktop with a 16 core CPU and two RTX 2080 Super GPUs. I am able to run training (I waited to verify that Epoch 0 completes successfully) when running the following command:
```
python run.py --adhoc --cfg conf/c02_selfplay/liars_sp.yaml \
    env.num_dice=1 \
    env.num_faces=4 \
    env.subgame_params.use_cfr=true \
    selfplay.cpu_gen_threads=16
```

When doing so I find that while the CPUs are 100% utilized, the GPUs are nearly unused. Specifically I see 100% CPU utilization on HTOP but I see 2-5% GPU utilization on one of my two GPUs when running nvidia-smi. I do see that the training code has allocated roughly 809MiB on the GPU, so I know that it is using it in some form.

I am new to reinforcement learning, but it would seem that I might be able to decrease training time by increasing my use of the GPU(s) during self play. The paper says "we use a single machine for training and up to 128 machines with 8 GPUs each for
data generation." and I see that lines 187-197 of rebel/cfvpy/selfplay.py seem to indicate that enabling the GPU might require not specifying anything for self.cfg.selfplay.cpu_gen_threads. When I remove selfplay.cpu_gen_threads=16 from the command above and comment out the same in rebel/conf/c02_selfplay/liars_sp.yaml it seems that the code cannot progress beyond warming up the replay buffer. Log output can be seen below. What might I be doing wrong? Thanks!

I0109 18:50:18 [selfplay:296] Train set size (forced): 25600
I0109 18:50:18 [selfplay:307] model size is 72964
I0109 18:50:18 [selfplay:311] Creating savedir: ckpt
I0109 18:50:18 [selfplay:319] warming up replay buffer: 0/1024
I0109 18:50:48 [selfplay:319] warming up replay buffer: 0/1024
I0109 18:51:18 [selfplay:319] warming up replay buffer: 0/1024
I0109 18:51:48 [selfplay:319] warming up replay buffer: 0/1024
I0109 18:52:18 [selfplay:319] warming up replay buffer: 0/1024
I0109 18:52:48 [selfplay:319] warming up replay buffer: 0/1024

**Comment by akhti** [comment]
>  and I see that lines 187-197 of rebel/cfvpy/selfplay.py seem to indicate that enabling the GPU might require not specifying anything for self.cfg.selfplay.cpu_gen_threads
The command uses CPUs for generation so that it works for people without a second GPU. To use GPU, add the following flags: `selfplay.cpu_gen_threads=0 selfplay.threads_per_gpu=8`. I'll update the docs.

A few more pointers on speeding up things for your setup.


A common practice is separate GPUs that are used for training and data generation. Specifically, Rebel reserves GPU0 for training and rest for data generation [0]. As a result, GPU 0 is doing almost nothing is it waits for enough data to be generated. [1] limits the amount of overfitting, i.e., the number of times we go through each example. 

You can change the code [0] start the range of act_devices from 0/add an option for that. This will make generation 2x faster.

Another way to make generation faster, is to use faster generation params. E.g. decrease number of CFR iters (pass `env. subgame_params.num_iters=32`) and use smaller net (pass `model.kwargs.n_hidden=64`).


[0] https://github.com/facebookresearch/rebel/blob/master/cfvpy/selfplay.py#L193 
[1] https://github.com/facebookresearch/rebel/blob/master/cfvpy/selfplay.py#L391
